compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/DataflowAnalyzer.java

```diff
@@ -996,19 +996,34 @@ public void visit(BLangTableConstructorExpr tableConstructorExpr) {
     private void checkForDuplicateKeys(BLangTableConstructorExpr tableConstructorExpr) {
         Set<Integer> keyHashSet = new HashSet<>();
         List<String> fieldNames = getFieldNames(tableConstructorExpr);
+        HashMap<Integer, List<BLangExpression>> keyValues = new HashMap<>();
         if (!fieldNames.isEmpty()) {
             for (BLangRecordLiteral literal : tableConstructorExpr.recordLiteralList) {
                 List<BLangExpression> keyArray = createKeyArray(literal, fieldNames);
                 int hashInt = generateHash(keyArray);
-                if (!keyHashSet.add(hashInt)) {
+                if (!keyHashSet.add(hashInt) && checkForKeyEquality(keyValues, keyArray, hashInt)) {
                     String fields = String.join(", ", fieldNames);
                     String values = keyArray.stream().map(Object::toString).collect(Collectors.joining(", "));
                     dlog.error(literal.pos, DiagnosticErrorCode.DUPLICATE_KEY_IN_TABLE_LITERAL, fields, values);
                 }
+                keyValues.put(hashInt, keyArray);
             }
         }
     }
 
+    private boolean checkForKeyEquality(HashMap<Integer, List<BLangExpression>> keyValues,
+                                        List<BLangExpression> keyArray, int hash) {
+        List<BLangExpression> existingExpList = keyValues.get(hash);
+        if (existingExpList.size() == keyArray.size()) {
+            boolean isEqual = true;
+            for (int i = 0; i < keyArray.size(); i++) {
+                isEqual = isEqual && equality(keyArray.get(i), existingExpList.get(i));
+            }
+            return isEqual;
+        }
+        return false;
```

Comment:
```suggestion
        List<BLangExpression> existingExpList = keyValues.get(hash);
        boolean isEqual = false;
        if (existingExpList.size() == keyArray.size()) {
            isEqual = true;
            for (int i = 0; i < keyArray.size(); i++) {
                isEqual = isEqual && equality(keyArray.get(i), existingExpList.get(i));
            }
        }
        return isEqual;
```

Reply:
Seems 1024 is never hit by codecov,

Reply:
Updated in https://github.com/ballerina-platform/ballerina-lang/pull/35130/commits/34c4d787ea2d75f7020af780491c470b7df32f25